### DISCUSSION TITLE
`@firebase/performance` needs some changes, to work as an ES module

### DISCUSSION THREAD

**Discussion by akauppi** [firebase/firebase-js-sdk#4636] [Q&A]
While other modules work with ES modules based browsers, `@firebase/performance` (0.900.17) needs some changes.

With ESM, one shouldn't do dynamic initialisation in the module body with classes that themselves are imported. The current code does this at places, causing errors like "Cannot access `ErrorFactory` before initialisation" to arise, at runtime.

Here are the places where I stabbed `node_modules/@firebase/performance/dist/index.esm2017.js` to get by them.

Note: Actual changes should be carefully though after by Firebase authors. I am merely listing the work-arounds. Preceding code is in the comments.

```
//const ERROR_FACTORY = new ErrorFactory(SERVICE, SERVICE_NAME, ERROR_DESCRIPTION_MAP);

const ERROR_FACTORY = {
  create(...args) {   // First time here. Replace ourselves (avoids calling 'new' at module load).
    const { create } = new ErrorFactory(SERVICE, SERVICE_NAME, ERROR_DESCRIPTION_MAP);
    ERROR_FACTORY.create = create;
    return create(...args);
  }
}
```

```
//const consoleLogger = new Logger(SERVICE_NAME);
//consoleLogger.logLevel = LogLevel.INFO;

let consoleLogger = {
  info: _consoleLoggerFirst('info')
};
function _consoleLoggerFirst(level) {
  return (...args) => {
    const logger = new Logger(SERVICE_NAME);
    logger.logLevel = LogLevel.INFO;

    consoleLogger= logger;
    logger[level](...args);
  }
}
```

```
//registerPerformance();
//registerVersion(name, version);

let registered;
function getPerformanceWrapper(...args) {
  if (!registered) {
    registerPerformance();
    registerVersion(name, version);
    registered = true;
  }
  return getPerformance(...args)
}
function initializePerformanceWrapper(...args) {
  if (!registered) {
    registerPerformance();
    registerVersion(name, version);
    registered = true;
  }
  return initializePerformance(...args)
}
```

```
//export { getPerformance, initializePerformance, trace };
export { getPerformanceWrapper as getPerformance, initializePerformanceWrapper as initializePerformance, trace };
```

*Edit*:

I complemented some of the snippets that had remain partial, and added a wrapper for `initializePerformanceWrapper` (shot myself in the foot by having changed to using it!).

**Comment by Feiyang1** [comment]
IIUC that can only happen if there is a circular dependency which is not the case for `ErrorFactory` which is imported from `@firebase/util`.

Did you actually encounter this error in your app?

**Reply by akauppi** [reply]
I did. The circular dependency maybe comes from that my code imports both `@firebase/app` and `@firebase/performance`. The latter imports `@firebase/app`, at least.

Hey, I have the repo here for you: https://github.com/akauppi/GroundLevel-firebase-es/tree/recamp-firebase-9x

`packages/app-deploy-ops/src/main.js`

**Reply by akauppi** [reply]
The problem remains in 0.900.19, where I get the following error without edits:

```
index.esm2017.js:77 Uncaught ReferenceError: Cannot access 's' before initialization
```

I understand this can be a particular use case and not a general concern.
